### PR TITLE
Retrieving underlay criteria configs from back-end

### PR DESCRIPTION
**Pull request by neelismail**

### CODE REVIEW

File: api/src/main/java/bio/terra/tanagra/app/controller/UnderlaysApiController.java [hunk 1]
```diff
@@ -60,12 +60,35 @@ public ResponseEntity<ApiListUnderlaysResponse> listUnderlays(
                     .map(UnderlaysApiController::convert)
                     .collect(Collectors.toList()))
             .nextPageToken(page.nextPageToken());
+
```
Comment (tjennison-google) [inline api/src/main/java/bio/terra/tanagra/app/controller/UnderlaysApiController.java:63]: Avoid including unrelated changes in PRs whenever possible.

File: api/src/main/java/bio/terra/tanagra/app/controller/UnderlaysApiController.java [hunk 2]
```diff
@@ -60,12 +60,35 @@ public ResponseEntity<ApiListUnderlaysResponse> listUnderlays(
...
     return ResponseEntity.ok(response);
   }
 
   private static ApiUnderlay convert(Underlay underlay) {
     List<String> entityNames =
         underlay.entities().keySet().stream().sorted().collect(Collectors.toList());
-    return new ApiUnderlay().name(underlay.name()).entityNames(entityNames);
+
+    String columns = "[" +
```
Comment (tjennison-google) [inline api/src/main/java/bio/terra/tanagra/app/controller/UnderlaysApiController.java:71]: There's a step missing here and maybe I wasn't clear enough when we discussed this before. This data shouldn't be hardcoded here, or anywhere else, in the code. Instead, it should be in the underlays (e.g. aou_synthetic.yaml) and read along with the rest of the information about the underlay like the entities. That path goes through underlay.proto, UnderlayConversion.java, and Underlay.java.

File: api/src/test/java/bio/terra/tanagra/app/controller/UnderlaysApiControllerTest.java [hunk 3]
```diff
@@ -24,13 +24,36 @@ public class UnderlaysApiControllerTest extends BaseSpringUnitTest {
...
+
   /** ApiUnderlay for the underlay specified by the nautical profile. */
   private static final ApiUnderlay NAUTICAL_API_UNDERLAY =
       new ApiUnderlay()
           .name(NauticalUnderlayUtils.NAUTICAL_UNDERLAY_NAME)
           .entityNames(
               ImmutableList.of(
-                  "boat_electric_anchors", "boat_engines", "boats", "reservations", "sailors"));
+                  "boat_electric_anchors", "boat_engines", "boats", "reservations", "sailors"))
+              .criteriaConfigs(criteriaConfigs);
```
Comment (tjennison-google) [inline api/src/test/java/bio/terra/tanagra/app/controller/UnderlaysApiControllerTest.java:56]: This looks too far indented to me. It should be on the same level as `.entityNames(` I think. I'm surprised the build process didn't complain.

File: ui/src/apiContext.ts [hunk 4]
```diff
@@ -4,12 +4,139 @@ import * as tanagra from "./tanagra-api";
 // TODO(tjennison): Figure out a more comprehensive solutions for faking APIs.
 class FakeUnderlaysApi {
   async listUnderlays(): Promise<tanagra.ListUnderlaysResponse> {
+    const columns = [
```
Comment (tjennison-google) [inline ui/src/apiContext.ts:7]: This should be a minimal configuration that satisfies the tests. You can check exactly what the tests need but I would expect something like 2 columns and one of each criteria type.

File: api/src/test/java/bio/terra/tanagra/app/controller/UnderlaysApiControllerTest.java [hunk 5]
```diff
@@ -24,13 +24,36 @@ public class UnderlaysApiControllerTest extends BaseSpringUnitTest {
 
   @Autowired private UnderlaysApiController controller;
 
+  // The static string variables below are only included here because they are required
```
Comment (tjennison-google) [inline api/src/test/java/bio/terra/tanagra/app/controller/UnderlaysApiControllerTest.java:27]: This should be super minimal since it's only for this one test. I would define a very short version as a constant and assert that it's passed through correctly in the `getUnderlay` test.

**Review by tjennison-google — CHANGES_REQUESTED** [review]

File: api/src/test/java/bio/terra/tanagra/service/underlay/NauticalUnderlayUtils.java [hunk 6]
```diff
@@ -35,6 +35,8 @@ public static bio.terra.tanagra.proto.underlay.Underlay loadNauticalUnderlayProt
   }
 
   public static final String NAUTICAL_UNDERLAY_NAME = "nautical_underlay";
+  public static final String NAUTICAL_CRITERIA_CONFIGS =
+      "[{\"type\":\"attribute\",\"title\":\"Age\",\"defaultName\":\"Age\",\"plugin\":{\"attribute\":\"age\"}]";
```
Comment (tjennison-google) [inline api/src/test/java/bio/terra/tanagra/service/underlay/NauticalUnderlayUtils.java:39]: See if you can use the `"""` string notation to avoid the escaped quotes.

File: api/src/test/resources/underlays/nautical.pbtext [hunk 7]
```diff
@@ -2,9 +2,10 @@
 # proto-message: Underlay
 # An Underlay prototext file describing a fictional relational domain of sailors, boats and
 # reservations for testing.
-# This should be euivalent to nautical.yaml
+# This should be equivalent to nautical.yaml
 
 name: "nautical_underlay"
+criteria_configs: "[{\"type\":\"attribute\",\"title\":\"Age\",\"defaultName\":\"Age\",\"plugin\":{\"attribute\":\"age\"}]"
```
Comment (tjennison-google) [inline api/src/test/resources/underlays/nautical.pbtext:8]: Use a single quoted string to avoid escaping the quotes inside.

File: api/src/test/resources/underlays/nautical.yaml [hunk 8]
```diff
@@ -3,6 +3,7 @@
 # This should be equivalent to nautical.pbtext.
 
 name: nautical_underlay
+criteriaConfigs: "[{\"type\":\"attribute\",\"title\":\"Age\",\"defaultName\":\"Age\",\"plugin\":{\"attribute\":\"age\"}]"
```
Comment (tjennison-google) [inline api/src/test/resources/underlays/nautical.yaml:6]: Use a single quoted string to avoid escaping the quotes inside.

File: lib/src/main/proto/bio/terra/tanagra/underlay/underlay.proto [hunk 9]
```diff
@@ -42,7 +42,10 @@ message Underlay {
 
   // The entity filters available on this underlay.
   repeated EntityFiltersSchema entity_filters_schemas = 8;
-
+
```
Comment (tjennison-google) [inline lib/src/main/proto/bio/terra/tanagra/underlay/underlay.proto:45]: Remove this whitespace change.

File: api/src/main/resources/underlays/aou_synthetic.yaml [hunk 10]
```diff
@@ -276,6 +276,7 @@ entities:
...
...id\",\"standard_concept\",\"concept_code\"],\"listChildren\":{\"entity\":\"ingredient\",\"idPath\":\"relationshipFilter.filter.binaryFilter.attributeValue\",\"filter\":{\"relationshipFilter\":{\"outerVariable\":\"ingredient\",\"newVariable\":\"brand\",\"newEntity\":\"brand\",\"filter\":{\"binaryFilter\":{\"attributeVariable\":{\"variable\":\"brand\",\"name\":\"concept_id\"},\"operator\":\"EQUALS\",\"attributeValue\":{\"int64Val\":0}}}}}}}]}},{\"type\":\"attribute\",\"title\":\"Ethnicity\",\"defaultName\":\"Contains Ethnicity Codes\",\"plugin\":{\"attribute\":\"ethnicity_concept_id\"}},{\"type\":\"attribute\",\"title\":\"Gender Identity\",\"defaultName\":\"Contains Gender Identity Codes\",\"plugin\":{\"attribute\":\"gender_concept_id\"}},{\"type\":\"attribute\",\"title\":\"Race\",\"defaultName\":\"Contains Race Codes\",\"plugin\":{\"attribute\":\"race_concept_id\"}},{\"type\":\"attribute\",\"title\":\"Sex Assigned at Birth\",\"defaultName\":\"Contains Sex Assigned at Birth Codes\",\"plugin\":{\"attribute\":\"sex_at_birth_concept_id\"}},{\"type\":\"attribute\",\"title\":\"Year at Birth\",\"defaultName\":\"Contains Year at Birth Values\",\"plugin\":{\"attribute\":\"year_of_birth\"}}]"
```
[the remark this answers is not in the document]
Comment (neelismail) [reply api/src/main/resources/underlays/aou_synthetic.yaml:279]: Done.

File: api/src/test/java/bio/terra/tanagra/service/underlay/NauticalUnderlayUtils.java (hunk 6 above)
Comment (neelismail) [reply api/src/test/java/bio/terra/tanagra/service/underlay/NauticalUnderlayUtils.java:39]: I tried using the `"""` notation but got the following complaint: `Text block literals are not supported at language level '11'`. Would escaped quotes be okay?

File: api/src/main/resources/underlays/aou_synthetic.yaml [hunk 11]
```diff
@@ -276,6 +276,8 @@ entities:
...
...ties":[{"name":"ingredient","selectable":true,"hierarchical":true},{"name":"brand","sourceConcepts":true,"attributes":["concept_name","concept_id","standard_concept","concept_code"],"listChildren":{"entity":"ingredient","idPath":"relationshipFilter.filter.binaryFilter.attributeValue","filter":{"relationshipFilter":{"outerVariable":"ingredient","newVariable":"brand","newEntity":"brand","filter":{"binaryFilter":{"attributeVariable":{"variable":"brand","name":"concept_id"},"operator":"EQUALS","attributeValue":{"int64Val":0}}}}}}}]}},{"type":"attribute","title":"Ethnicity","defaultName":"Contains Ethnicity Codes","plugin":{"attribute":"ethnicity_concept_id"}},{"type":"attribute","title":"Gender Identity","defaultName":"Contains Gender Identity Codes","plugin":{"attribute":"gender_concept_id"}},{"type":"attribute","title":"Race","defaultName":"Contains Race Codes","plugin":{"attribute":"race_concept_id"}},{"type":"attribute","title":"Sex Assigned at Birth","defaultName":"Contains Sex Assigned at Birth Codes","plugin":{"attribute":"sex_at_birth_concept_id"}},{"type":"attribute","title":"Year at Birth","defaultName":"Contains Year at Birth Values","plugin":{"attribute":"year_of_birth"}}]"
```
Comment (tjennison-google) [inline api/src/main/resources/underlays/aou_synthetic.yaml:280]: Reformat this sensibly, since we'll need to be able to edit it.

Comment (tjennison-google) [reply api/src/main/resources/underlays/aou_synthetic.yaml]: Whitespace is allowed in JSON so that doesn't cause an issue, does it? There are additional YAML options for how the whitespace is treated that could be helpful in that case.

**Review by tjennison-google — APPROVED** [review]
